cpp/include/graph.hpp

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <utilities/error_utils.h>
+#include "utilities/graph_utils.cuh"
+
+namespace cugraph {
+namespace experimental {
+
+typedef enum prop_type{PROP_UNDEF, PROP_FALSE, PROP_TRUE} PropType;
+
+  // TODO:
+  //    New idea... get rid of EdgeList and AdjList
+  //    Create three classes:
+  //        GraphCOO<VT,WT>
+  //        GraphCSR<VT,WT>
+  //        GraphCSC<VT,WT>
+  //
+  //    Then checks to be sure we're using the correct graph type can be done at compile time.
+  //
+
+struct GraphProperties {
+  bool directed;
+  bool weighted;
+  bool multigraph;
+  bool bipartite;
+  bool tree;
+  PropType has_negative_edges;
+  GraphProperties() : directed(false), weighted(false), multigraph(false), bipartite(false), tree(false), has_negative_edges(PROP_UNDEF){}
+};
+
+/**
+ * @Synopsis    A graph stored in COO (COOrdinate) format.
+ *
+ * @tparam VT   Type of vertex (defaults to int)
+ * @tparam WT   Type of weight (defaults to float)
+ */
+template <typename VT = int, typename WT = float>
+class GraphCOO {
+public:
+  VT const *src_indices;   ///< rowInd
+  VT const *dst_indices;   ///< colInd
+  WT const *edge_data;     ///< edge weight
+
+  GraphProperties          prop;
```

Comment:
graph prop, V and E are shared for all graph formats, they could be in a parent `Graph` class.

Reply:
I thought about that, could definitely add a parent graph class and reduce some duplication here.

Reply:
addressed in https://github.com/rapidsai/cugraph/pull/722/commits/09adaa80b619c811ced0ab08f977237aebca6c5d